docs/hugo/content/design/ADR-2023-09-Skipping-Properties/_index.md

```diff
@@ -0,0 +1,418 @@
+
+[v4.Person|FullName string;FamilyName string; KnownAs string|Properties PropertyBag]
+[v4.Person]-.-[v5/compat.Address{bg:cornsilk}|Label multilineString|Properties PropertyBag]
+
+[v5.Person|FullName string;FamilyName string; KnownAs string;ResidentialAddress v5.Address|Properties PropertyBag]
+[v5.Person]->[v5.Address|Street string; Suburb string; City string; Country string|Properties PropertyBag]
+
+-->
+
+{{< figure src="./option-3-class.png" >}}
+
+The invariant we need to maintain here is that the **v5/compat** version of `Address` needs to have the same exact shape as the **v3** version of `Address`.
+
+Pros:
+* Addresses the central inconsistency problem, ensuring we always serialize the same shape into intermediate property bags
+* Preserves the serialization format for existing resource versions (e.g. **v3** and **v4**)
+* Does not change the object model for released code
+* Isolates types introduced for compatibility purposes in a dedicated subpackage, avoiding the problem of circular dependencies.
+
+Cons:
+* Still reverses the direction of our existing conversion contract, which has all the conversion code resident on the earlier version (e.g. **v4**). Moving this code to **v5** in order to avoid changes to **v4** would require serious rework in our code generator.
+* When there are multi-generational skips, other questions arise: Do we introduce the new type into every intermediate version, only the last, or both first and last? Each approach has advantages and drawbacks.
+
+### Option 4: Forward conversion with compatibility type
+
+As for Option 3, but we create the compatibility type in a subpackage of **v4**, replicating the shape of the **v3** `Address`. When we generate code to read `ResidentialAddress` from the PropertyBag on a **v4** Person, we deserialize into a **v4/compat** `Address` (with an identical shape to **v3**), and then convert that to a **v5** `Address` for use on the **v5** `Person``.
+
+<!-- yuml.me
+
+[v3.Person|FullName string;FamilyName string; KnownAs string; ResidentialAddress v3.Address|Properties PropertyBag]
+[v3.Person]->[v3.Address|Label multilineString|Properties PropertyBag]
+
+[v4.Person|FullName string;FamilyName string; KnownAs string|Properties PropertyBag]
+[v4.Person]-.-[v4/compat.Address{bg:cornsilk}|Label multilineString|Properties PropertyBag]
+
+[v5.Person|FullName string;FamilyName string; KnownAs string;ResidentialAddress v5.Address|Properties PropertyBag]
+[v5.Person]->[v5.Address|Street string; Suburb string; City string; Country string|Properties PropertyBag]
+
+-->
+
+{{< figure src="./option-4-class.png" >}}
+
+Conversion from **v4** to **v5** would work as follows:
+
+* Check the property bag for a `ResidentialAddress` property.
+* If found, deserialize it as a **v4/compat** `Address`.
+* Recursively invoke a conversion method to convert the **v4/compat** `Address` to a **v5** `Address` (this mechanism already exists in our property conversion framework).
```

Comment:
As I alluded to before, once the `v4.Address` (or `v4/compat.Address`) type exists the conversion logic for going to/from `v5` would live on that type, in the same way that `v4` references `v5` for doing conversion today, so I don't understand the need for the subpackage.

Am I correct that the only difference between Option 2 and Option 4 is the subpackage?